File: 2026/dp/max_dot_product_1458.py
class Solution:
    def __init__(self):
        self.m = 0
        self.n = 0
        self.t = [[-int(1e9)] * 501 for _ in range(501)]
        
    def solve(self, nums1: List[int], nums2: List[int], i: int, j: int) -> int:
        if i == self.m or j == self.n:
            return -100000000  # large negative number
        if self.t[i][j] != -int(1e9):
            return self.t[i][j] 
        val = nums1[i] * nums2[j]
        take_i_j = self.solve(nums1, nums2, i + 1, j+1) + val
        take_i = self.solve(nums1, nums2, i, j+1)
        take_j = self.solve(nums1, nums2, i+1, j)
        self.t[i][j] = max(val, take_i_j, take_i, take_j)
        return self.t[i][j]

    def maxDotProduct(self, nums1: List[int], nums2: List[int]) -> int:
        self.m = len(nums1)
        self.n = len(nums2)
        for i in range(501):
            for j in range(501):
                self.t[i][j] = -int(1e9)
        return self.solve(nums1, nums2, 0, 0)

File: 2026/dp/max_dot_product_1458.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
 */
var maxDotProduct = function (nums1, nums2) {
    let m = nums1.length;
    let n = nums2.length;
    const t = Array.from({ length: 501 }, () => Array(501).fill(-1e9));
    const solve = (i, j) => {
        if (i === m || j === n) {
            return -1e9;
        }
        if (t[i][j] !== -1e9) {
            return t[i][j];
        }
        const val = nums1[i] * nums2[j];
        const take_i_j = solve(i + 1, j + 1) + val;
        const take_i = solve(i, j + 1);
        const take_j = solve(i + 1, j);
        return t[i][j] = Math.max(val, take_i_j, take_i, take_j);
    };
    return solve(0, 0);
};